File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        if len(nums) == 1:
            return [nums.copy()]
        
        for num in nums:
            val = nums.pop(0)
            perms = self.permute(nums)
            for perm in perms:
                perm.append(val)
            res.extend(perms)
            nums.append(val)
        return res
                

File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> subset = new ArrayList<>();
        backtrack(res, subset, 0, nums);
        return res;
    }
    public void backtrack(List<List<Integer>> res,
                        List<Integer> subset,
                        int start, 
                        int[] nums){
                            if(subset.size() == nums.length){
                                res.add(new ArrayList<>(subset));
                                return;
                            }else{
                                for(int i = 0; i < nums.length; i++){
                                    if(!subset.contains(nums[i])){
                                        subset.add(nums[i]);
                                        backtrack(res, subset, i + 1, nums);
                                        subset.remove(subset.size() - 1);
                                    }
                                }
                            }
                        }
}